p4-16/spec/P4-16-spec.mdk

```diff
@@ -2363,6 +2363,85 @@ struct Parsed_headers {
 }
 ~ End P4Example
 
+### Structure overlays { #sec-overlays }
+
+An overlay allows the access to fields or subfields of a `struct` or
+`header` using alternative names.  An overlay is constructed by a
+special `overlay` function, that establishes a connection between the
+fields of the source and destination structure.
+
+In the following example we have two structure types: `Addresses` and
+`Fields`.  The overlay named `sourceQuads` is given a structure of
+type `Addresses` and allows accesses to its fields through a structure
+of type `Fields`.  The first four fields of `Fields` are individual
+bytes of `Addresses.source`, and the field `Fields.all` is the
+concatenation of `Source.source` and `Source.destination`.
+
+A variable of type `Addresses` named `a` is instantiated, and an
+overlay named `q` is created using the `sourceQuads` overlay.  Now
+fields of the `a` structure can be accessed read-write using fields of
+the `q` structure.
+
+~ Begin P4Example
+struct Addresses {
+    bit<32> source;
+    bit<32> destination;
+}
+
+struct Fields {
+    bit<8> q0;
+    bit<8> q1;
+    bit<8> q2;
+    bit<8> q3;
+    bit<64> all;
+}
+
+overlay Fields sourceQuads(Addresses add) {
+   q0 = add.source[31:24];
+   q1 = add.source[23:16];
+   q2 = add.source[15:8];
+   q3 = add.source[7:0];
+   all = add.source ++ add.destination;
+}
+
+Addresses a = { 0x01234567, 0x89ABCDEF };
+Fields    q = sourceQuads(a);  // q is { 0x01, 0x23, 0x45, 0x67, 0x0123456789ABCDEF };
```

Comment:
This appears to be not an assignment like most others in P4_16, in that by doing this it actually causes q to be overlayed with a.  Is that correct?

If so, is it allowed later to do another assignment to 'q' with a different overlay, or the same overlay with a different parameter other than 'a'?  If that is allowed, is it effectively changing what 'q' "points to" at that time?

Reply:
The example overlay sourceQuads does not meet the restrictions described later for the overlay being a left-value.  Should it still be allowed to assign values to fields of q, as is done immediately following in this example, if the overlay does not satisfy those restrictions?

Is it permitted to "stack" overlays?  That is, after defining q as shown above, can I define an overlay 'overlayA' that take type 'Fields' as an input parameter, and returns some other type T1, and then do 'T1 x = overlayA(q);', and now assignments to fields of x can modify fields of both q and a?  Is there any limit imposed by the feature on this?

What if I do the assignment to q above, then the one to 'x' in the previous paragraph, and then I assign 'a = newOverlay(x);'?